README.md
[![Build Status](https://travis-ci.com/BOPOHOB/meval-rs.svg?branch=master)](https://travis-ci.org/BOPOHOB/meval-rs)

# meval

Just usual [meval](https://github.com/rekka/meval-rs) but with time computation future

## Documentation

- [Original meval full API documentation](https://docs.rs/meval)
- The difference between it is additional function Expr::eval_with_unit with return additional information about secons degree in a diven expression (see example),

## Simple examples

```rust
extern crate meval;

fn main() {
    let r = meval::eval_str("2:06:35.1 / 42.195 / 3:00").unwrap();

    println!("2:06:35.1 / 42.195 / 3:00 = {}", r);
}
```

If you need to know the measute of result value

```rust
extern crate meval;

use std::str::FromStr;
use meval::{ Expr, Error };

pub fn eval_str<S: AsRef<str>>(expr: S) -> Result<(f64, f64), Error> {
    let expr = Expr::from_str(expr.as_ref())?;
    expr.eval_with_unit()
}

pub fn sweet_eval<S: AsRef<str>>(expr: S) -> Result<String, Error> {
    let (result, degree) = eval_str(expr)?;

    let rounded = (result * 1000.0).round() / 1000.0;
    if degree == 1. {
        let sign = if rounded > 0.0 { "" } else { "-" };
        let mut seconds = rounded.abs();
        let hours = (seconds / 3600.0).floor() as i64;
        seconds -= (hours * 3600) as f64;
        let minutes = (seconds / 60.0).floor() as i64;
        seconds -= (minutes * 60) as f64;
        Ok(format!("{}{}:{:02}:{:07.4}", sign, hours, minutes, seconds))
    } else {
        Ok(format!("{}", rounded))
    }
}

fn main() {
    let pase = sweet_eval("42.195 * 3:08").unwrap();
    println!("The marathon pase time with pase 3:08 per kilometer is {}", pase);

    let (result, degree) = eval_str("10000 * 2 / 1:17^2").unwrap();
    assert_eq!(degree, -2.0);
    println!("The acceleration of something is {}m/c^2", result);
}

```


CHANGELOG.md
# 0.2.1
- `:` operator and `eval_with_unit` function

# 0.2

- `bind` and `bind_with_context` functions return unboxed closures
  (requires Rust 1.26).

- [#17](https://github.com/rekka/meval-rs/pull/17)

    - Implement `bind4`, `bind4_with_context`, `bind5`,
      `bind5_with_context` and `bindn`.

    - Bugfix: Serialization from JSON now works as expected.

- Drop support for serialization from non-self-describing formats.

# 0.1.1

- Implement `Default` for `Context`

# 0.1.0

- Support `serde-1.0.0`

# 0.0.9

- Bug #13: failed build with `serde` feature disabled

# 0.0.8

- added serde deserialization
- api change: `Expr::eval` is now `Expr::eval_with_context`
- `Expr` implements `FromStr`: use it as "1 + 2".parse::<Expr>()
- added `de::as_f64` for convenient deserialization
- `Context` is now `Clone`


src/de.rs
//! Deserialization utilities.
use super::Expr;
use serde::de;
use serde::de::Error;
use serde::Deserialize;

/// Deserialize into [`Expr`](../struct.Expr.html) and then evaluate using `Expr::eval`.
///
/// # Example
///
/// ```rust
/// #[macro_use]
/// extern crate toml;
/// extern crate meval;
/// extern crate serde;
/// use serde::Deserialize;
/// use meval::{Expr, Context};
///
/// #[derive(Deserialize)]
/// struct Foo {
///     #[serde(deserialize_with = "meval::de::as_f64")]
///     x: f64,
/// }
///
/// fn main() {
///     let foo: Foo = toml::from_str(r#" x = "cos(1.)" "#).unwrap();
///     assert_eq!(foo.x, 1f64.cos());
///
///     let foo: Result<Foo, _> = toml::from_str(r#" x = "cos(x)" "#);
///     assert!(foo.is_err());
/// }
/// ```
///
/// See [crate root](../index.html#deserialization) for another example.
pub fn as_f64<'de, D: de::Deserializer<'de>>(deserializer: D) -> Result<f64, D::Error> {
    Expr::deserialize(deserializer)?
        .eval()
        .map_err(D::Error::custom)
}


src/tokenizer.rs
//! Tokenizer that converts a mathematical expression in a string form into a series of `Token`s.
//!
//! The underlying parser is build using the [nom] parser combinator crate.
//!
//! The parser should tokenize only well-formed expressions.
//!
//! [nom]: https://crates.io/crates/nom
use nom::{multispace, slice_to_offsets, IResult, Needed};
use std;
use std::fmt;
use std::str::from_utf8;
#[cfg(feature = "serde")]
extern crate serde;
#[cfg(feature = "serde")]
use serde::{Serialize, Deserialize};

/// An error reported by the parser.
#[derive(Debug, Clone, PartialEq)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub enum ParseError {
    /// A token that is not allowed at the given location (contains the location of the offending
    /// character in the source string).
    UnexpectedToken(usize),
    /// Missing right parentheses at the end of the source string (contains the number of missing
    /// parens).
    MissingRParen(i32),
    /// Missing operator or function argument at the end of the expression.
    MissingArgument,
}

impl fmt::Display for ParseError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match *self {
            ParseError::UnexpectedToken(i) => write!(f, "Unexpected token at byte {}.", i),
            ParseError::MissingRParen(i) => write!(
                f,
                "Missing {} right parenthes{}.",
                i,
                if i == 1 { "is" } else { "es" }
            ),
            ParseError::MissingArgument => write!(f, "Missing argument at the end of expression."),
        }
    }
}

impl std::error::Error for ParseError { }

/// Mathematical operations.
#[derive(Debug, PartialEq, Clone, Copy)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub enum Operation {
    Plus,
    Minus,
    Times,
    Div,
    Rem,
    Pow,
    Fact,
    Colon,
}

/// Expression tokens.
#[derive(Debug, PartialEq, Clone)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub enum Token {
    /// Binary operation.
    Binary(Operation),
    /// Unary operation.
    Unary(Operation),

    /// Left parenthesis.
    LParen,
    /// Right parenthesis.
    RParen,
    /// Comma: function argument separator
    Comma,

    /// A number.
    Number(f64),
    /// A variable.
    Var(String),
    /// A function with name and number of arguments.
    Func(String, Option<usize>),
}

named!(
    binop<Token>,
    alt!(
        chain!(tag!("+"), || Token::Binary(Operation::Plus))
            | chain!(tag!("-"), || Token::Binary(Operation::Minus))
            | chain!(tag!("*"), || Token::Binary(Operation::Times))
            | chain!(tag!("/"), || Token::Binary(Operation::Div))
            | chain!(tag!("%"), || Token::Binary(Operation::Rem))
            | chain!(tag!("^"), || Token::Binary(Operation::Pow))
            | chain!(tag!(":"), || Token::Binary(Operation::Colon))
    )
);

named!(
    negpos<Token>,
    alt!(
        chain!(tag!("+"), || Token::Unary(Operation::Plus))
            | chain!(tag!("-"), || Token::Unary(Operation::Minus))
    )
);

named!(
    fact<Token>,
    chain!(tag!("!"), || Token::Unary(Operation::Fact))
);
named!(lparen<Token>, chain!(tag!("("), || Token::LParen));
named!(rparen<Token>, chain!(tag!(")"), || Token::RParen));
named!(comma<Token>, chain!(tag!(","), || Token::Comma));

/// Parse an identifier:
///
/// Must start with a letter or an underscore, can be followed by letters, digits or underscores.
fn ident(input: &[u8]) -> IResult<&[u8], &[u8]> {
    use nom::Err::*;
    use nom::IResult::*;
    use nom::{ErrorKind};

    // first character must be 'a'..='z' | 'A'..='Z' | '_'
    match input.first().cloned() {
        Some(b'a'..=b'z') | Some(b'A'..=b'Z') | Some(b'_') => {
            let n = input
                .iter()
                .skip(1)
                .take_while(|&&c| match c {
                    b'a'..=b'z' | b'A'..=b'Z' | b'_' | b'0'..=b'9' => true,
                    _ => false,
                })
                .count();
            let (parsed, rest) = input.split_at(n + 1);
            Done(rest, parsed)
        }
        None => Incomplete(Needed::Size(1)),
        _ => Error(Position(ErrorKind::Custom(0), input)),
    }
}

named!(
    var<Token>,
    map!(map_res!(complete!(ident), from_utf8), |s: &str| Token::Var(
        s.into()
    ))
);

named!(
    func<Token>,
    map!(
        map_res!(
            terminated!(
                complete!(ident),
                preceded!(opt!(multispace), complete!(tag!("(")))
            ),
            from_utf8
        ),
        |s: &str| Token::Func(s.into(), None)
    )
);

/// Matches one or more digit characters `0`...`9`.
///
/// Never returns `nom::IResult::Incomplete`.
///
/// Fix of IMHO broken `nom::digit`, which parses an empty string successfully.
fn digit_complete(input: &[u8]) -> IResult<&[u8], &[u8]> {
    use nom::Err::*;
    use nom::IResult::*;
    use nom::{is_digit, ErrorKind};

    let n = input.iter().take_while(|&&c| is_digit(c)).count();
    if n > 0 {
        let (parsed, rest) = input.split_at(n);
        Done(rest, parsed)
    } else {
        Error(Position(ErrorKind::Digit, input))
    }
}

named!(
    float<usize>,
    chain!(
        a: digit_complete ~
        b: complete!(chain!(tag!(".") ~ d: digit_complete?,
                            ||{1 + d.map(|s| s.len()).unwrap_or(0)}))? ~
        e: complete!(exp),
        ||{a.len() + b.unwrap_or(0) + e.unwrap_or(0)}
    )
);

/// Parser that matches the exponential part of a float. If the `input[0] == 'e' | 'E'` then at
/// least one digit must match.
fn exp(input: &[u8]) -> IResult<&[u8], Option<usize>> {
    use nom::IResult::*;
    match alt!(input, tag!("e") | tag!("E")) {
        Incomplete(_) | Error(_) => Done(input, None),
        Done(i, _) => match chain!(i, s: alt!(tag!("+") | tag!("-"))? ~
                   e: digit_complete,
                ||{Some(1 + s.map(|s| s.len()).unwrap_or(0) + e.len())})
        {
            Incomplete(Needed::Size(i)) => Incomplete(Needed::Size(i + 1)),
            o => o,
        },
    }
}

fn number(input: &[u8]) -> IResult<&[u8], Token> {
    use nom::Err;
    use nom::ErrorKind;
    use nom::IResult::*;
    use std::str::FromStr;

    match float(input) {
        Done(rest, l) => {
            // it should be safe to call unwrap here instead of the error checking, since
            // `float` should match only well-formed numbers
            from_utf8(&input[..l])
                .ok()
                .and_then(|s| f64::from_str(s).ok())
                .map_or(Error(Err::Position(ErrorKind::Custom(0), input)), |f| {
                    Done(rest, Token::Number(f))
                })
        }
        Error(e) => Error(e),
        Incomplete(n) => Incomplete(n),
    }
}

named!(
    lexpr<Token>,
    delimited!(
        opt!(multispace),
        alt!(number | func | var | negpos | lparen),
        opt!(multispace)
    )
);
named!(
    after_rexpr<Token>,
    delimited!(
        opt!(multispace),
        alt!(fact | binop | rparen),
        opt!(multispace)
    )
);
named!(
    after_rexpr_no_paren<Token>,
    delimited!(opt!(multispace), alt!(fact | binop), opt!(multispace))
);
named!(
    after_rexpr_comma<Token>,
    delimited!(
        opt!(multispace),
        alt!(fact | binop | rparen | comma),
        opt!(multispace)
    )
);

#[derive(Debug, Clone, Copy)]
enum TokenizerState {
    // accept any token that is an expression from the left: var, num, (, negpos
    LExpr,
    // accept any token that needs an expression on the left: fact, binop, ), comma
    AfterRExpr,
}

#[derive(Debug, Clone, Copy)]
enum ParenState {
    Subexpr,
    Func,
}

/// Tokenize a given mathematical expression.
///
/// The parser should return `Ok` only if the expression is well-formed.
///
/// # Failure
///
/// Returns `Err` if the expression is not well-formed.
pub fn tokenize<S: AsRef<str>>(input: S) -> Result<Vec<Token>, ParseError> {
    use self::TokenizerState::*;
    use nom::Err;
    use nom::IResult::*;
    let mut state = LExpr;
    // number of function arguments left
    let mut paren_stack = vec![];

    let mut res = vec![];

    let input = input.as_ref().as_bytes();
    let mut s = input;

    while !s.is_empty() {
        let r = match (state, paren_stack.last()) {
            (LExpr, _) => lexpr(s),
            (AfterRExpr, None) => after_rexpr_no_paren(s),
            (AfterRExpr, Some(&ParenState::Subexpr)) => after_rexpr(s),
            (AfterRExpr, Some(&ParenState::Func)) => after_rexpr_comma(s),
        };

        match r {
            Done(rest, t) => {
                match t {
                    Token::LParen => {
                        paren_stack.push(ParenState::Subexpr);
                    }
                    Token::Func(..) => {
                        paren_stack.push(ParenState::Func);
                    }
                    Token::RParen => {
                        paren_stack.pop().expect("The paren_stack is empty!");
                    }
                    Token::Var(_) | Token::Number(_) => {
                        state = AfterRExpr;
                    }
                    Token::Binary(_) | Token::Comma => {
                        state = LExpr;
                    }
                    _ => {}
                }
                res.push(t);
                s = rest;
            }
            Error(Err::Position(_, p)) => {
                let (i, _) = slice_to_offsets(input, p);
                return Err(ParseError::UnexpectedToken(i));
            }
            _ => {
                panic!(
                    "Unexpected parse result when parsing `{}` at `{}`: {:?}",
                    String::from_utf8_lossy(input),
                    String::from_utf8_lossy(s),
                    r
                );
            }
        }
    }

    match state {
        LExpr => Err(ParseError::MissingArgument),
        _ if !paren_stack.is_empty() => Err(ParseError::MissingRParen(paren_stack.len() as i32)),
        _ => Ok(res),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use super::{binop, func, number, var};
    use nom::Err::*;
    use nom::ErrorKind::*;
    use nom::IResult;

    #[test]
    fn it_works() {
        assert_eq!(
            binop(b"+"),
            IResult::Done(&b""[..], Token::Binary(Operation::Plus))
        );
        assert_eq!(
            number(b"32143"),
            IResult::Done(&b""[..], Token::Number(32143f64))
        );
        assert_eq!(
            var(b"abc"),
            IResult::Done(&b""[..], Token::Var("abc".into()))
        );
        assert_eq!(
            func(b"abc("),
            IResult::Done(&b""[..], Token::Func("abc".into(), None))
        );
        assert_eq!(
            func(b"abc ("),
            IResult::Done(&b""[..], Token::Func("abc".into(), None))
        );
    }

    #[test]
    fn test_var() {
        for &s in ["abc", "U0", "_034", "a_be45EA", "aAzZ_"].iter() {
            assert_eq!(
                var(s.as_bytes()),
                IResult::Done(&b""[..], Token::Var(s.into()))
            );
        }

        assert_eq!(var(b""), IResult::Error(Position(Complete, &b""[..])));
        assert_eq!(var(b"0"), IResult::Error(Position(Custom(0), &b"0"[..])));
    }

    #[test]
    fn test_func() {
        for &s in ["abc(", "u0(", "_034 (", "A_be45EA  ("].iter() {
            assert_eq!(
                func(s.as_bytes()),
                IResult::Done(
                    &b""[..],
                    Token::Func((&s[0..s.len() - 1]).trim().into(), None)
                )
            );
        }

        assert_eq!(func(b""), IResult::Error(Position(Complete, &b""[..])));
        assert_eq!(func(b"("), IResult::Error(Position(Custom(0), &b"("[..])));
        assert_eq!(func(b"0("), IResult::Error(Position(Custom(0), &b"0("[..])));
    }

    #[test]
    fn test_number() {
        assert_eq!(
            number(b"32143"),
            IResult::Done(&b""[..], Token::Number(32143f64))
        );
        assert_eq!(
            number(b"2."),
            IResult::Done(&b""[..], Token::Number(2.0f64))
        );
        assert_eq!(
            number(b"32143.25"),
            IResult::Done(&b""[..], Token::Number(32143.25f64))
        );
        assert_eq!(
            number(b"0.125e9"),
            IResult::Done(&b""[..], Token::Number(0.125e9f64))
        );
        assert_eq!(
            number(b"20.5E-3"),
            IResult::Done(&b""[..], Token::Number(20.5E-3f64))
        );
        assert_eq!(
            number(b"123423e+50"),
            IResult::Done(&b""[..], Token::Number(123423e+50f64))
        );

        assert_eq!(number(b""), IResult::Error(Position(Digit, &b""[..])));
        assert_eq!(number(b".2"), IResult::Error(Position(Digit, &b".2"[..])));
        assert_eq!(number(b"+"), IResult::Error(Position(Digit, &b"+"[..])));
        assert_eq!(number(b"e"), IResult::Error(Position(Digit, &b"e"[..])));
        assert_eq!(number(b"1E"), IResult::Error(Position(Complete, &b"E"[..])));
        assert_eq!(number(b"1e+"), IResult::Error(Position(Digit, &b""[..])));
    }

    #[test]
    fn test_tokenize() {
        use super::Operation::*;
        use super::Token::*;

        assert_eq!(tokenize("a"), Ok(vec![Var("a".into())]));

        assert_eq!(
            tokenize("2 +(3--2) "),
            Ok(vec![
                Number(2f64),
                Binary(Plus),
                LParen,
                Number(3f64),
                Binary(Minus),
                Unary(Minus),
                Number(2f64),
                RParen
            ])
        );

        assert_eq!(
            tokenize("-2^ ab0 *12 - C_0"),
            Ok(vec![
                Unary(Minus),
                Number(2f64),
                Binary(Pow),
                Var("ab0".into()),
                Binary(Times),
                Number(12f64),
                Binary(Minus),
                Var("C_0".into()),
            ])
        );

        assert_eq!(
            tokenize("45:00/1:24*1.981"),
            Ok(vec![
                Number(45f64),
                Binary(Colon),
                Number(0f64),
                Binary(Div),
                Number(1f64),
                Binary(Colon),
                Number(24f64),
                Binary(Times),
                Number(1.981),
            ])
        );

        assert_eq!(
            tokenize("-sin(pi * 3)^ cos(2) / Func2(x, f(y), z) * _buildIN(y)"),
            Ok(vec![
                Unary(Minus),
                Func("sin".into(), None),
                Var("pi".into()),
                Binary(Times),
                Number(3f64),
                RParen,
                Binary(Pow),
                Func("cos".into(), None),
                Number(2f64),
                RParen,
                Binary(Div),
                Func("Func2".into(), None),
                Var("x".into()),
                Comma,
                Func("f".into(), None),
                Var("y".into()),
                RParen,
                Comma,
                Var("z".into()),
                RParen,
                Binary(Times),
                Func("_buildIN".into(), None),
                Var("y".into()),
                RParen,
            ])
        );

        assert_eq!(
            tokenize("2 % 3"),
            Ok(vec![Number(2f64), Binary(Rem), Number(3f64)])
        );

        assert_eq!(
            tokenize("1 + 3! + 1"),
            Ok(vec![
                Number(1f64),
                Binary(Plus),
                Number(3f64),
                Unary(Fact),
                Binary(Plus),
                Number(1f64)
            ])
        );

        assert_eq!(tokenize("!3"), Err(ParseError::UnexpectedToken(0)));

        assert_eq!(tokenize("()"), Err(ParseError::UnexpectedToken(1)));

        assert_eq!(tokenize(""), Err(ParseError::MissingArgument));
        assert_eq!(tokenize("2)"), Err(ParseError::UnexpectedToken(1)));
        assert_eq!(tokenize("2^"), Err(ParseError::MissingArgument));
        assert_eq!(tokenize("(((2)"), Err(ParseError::MissingRParen(2)));
        assert_eq!(tokenize("f(2,)"), Err(ParseError::UnexpectedToken(4)));
        assert_eq!(tokenize("f(,2)"), Err(ParseError::UnexpectedToken(2)));
    }
}


src/lib.rs
//! This [Rust] crate provides a simple math expression parsing and evaluation. Its main goal is to
//! be convenient to use, while allowing for some flexibility. Currently works only with `f64`
//! types. A typical use case is the configuration of numerical computations in
//! Rust, think initial data and boundary conditions, via config files or command line arguments.
//!
//! # Documentation
//!
//! - [Full API documentation](https://docs.rs/meval)
//!
//! # Installation
//!
//! Simply add the corresponding entry to your `Cargo.toml` dependency list:
//!
//! ```toml
//! [dependencies]
//! meval = "0.2"
//! ```
//!
//! and add this to your crate root:
//!
//! ```rust
//! extern crate meval;
//! ```
//!
//!  **Requires Rust 1.26.**
//!
//! # Simple examples
//!
//! ```rust
//! extern crate meval;
//!
//! fn main() {
//!     let r = meval::eval_str("1 + 2").unwrap();
//!
//!     println!("1 + 2 = {}", r);
//! }
//! ```
//!
//! Need to define a Rust function from an expression? No problem, use [`Expr`][Expr]
//! for this and more:
//!
//! ```rust
//! extern crate meval;
//!
//! fn main() {
//!     let expr: meval::Expr = "sin(pi * x)".parse().unwrap();
//!     let func = expr.bind("x").unwrap();
//!
//!     let vs: Vec<_> = (0..100+1).map(|i| func(i as f64 / 100.)).collect();
//!
//!     println!("sin(pi * x), 0 <= x <= 1: {:?}", vs);
//! }
//! ```
//!
//! Custom constants and functions? Define a [`Context`][Context]!
//!
//! ```rust
//! use meval::{Expr, Context};
//!
//! let y = 1.;
//! let expr: Expr = "phi(-2 * zeta + x)".parse().unwrap();
//!
//! // create a context with function definitions and variables
//! let mut ctx = Context::new(); // built-ins
//! ctx.func("phi", |x| x + y)
//!    .var("zeta", -1.);
//! // bind function with a custom context
//! let func = expr.bind_with_context(ctx, "x").unwrap();
//! assert_eq!(func(2.), -2. * -1. + 2. + 1.);
//! ```
//!
//! For functions of 2, 3, and N variables use `Context::func2`, `Context::func3` and
//! `Context::funcn`,
//! respectively. See [`Context`][Context] for more options.
//!
//! If you need a custom function depending on mutable parameters, you will need to use a
//! [`Cell`](https://doc.rust-lang.org/stable/std/cell/struct.Cell.html):
//!
//! ```rust
//! use std::cell::Cell;
//! use meval::{Expr, Context};
//! let y = Cell::new(0.);
//! let expr: Expr = "phi(x)".parse().unwrap();
//!
//! let mut ctx = Context::empty(); // no built-ins
//! ctx.func("phi", |x| x + y.get());
//!
//! let func = expr.bind_with_context(ctx, "x").unwrap();
//! assert_eq!(func(2.), 2.);
//! y.set(3.);
//! assert_eq!(func(2.), 5.);
//! ```
//!
//! # Supported expressions
//!
//! `meval` supports basic mathematical operations on floating point numbers:
//!
//! - binary operators: `+`, `-`, `*`, `/`, `%` (remainder), `^` (power)
//! - unary operators: `+`, `-`
//!
//! It supports custom variables and functions like `x`, `weight`, `C_0`, `f(1)`, etc. A variable
//! or function name must start with `[a-zA-Z_]` and can contain only `[a-zA-Z0-9_]`. Custom
//! functions with a variable number of arguments are also supported.
//!
//! Build-ins (given by the context `Context::new()` and when no context provided) currently
//! supported:
//!
//! - functions implemented using functions of the same name in [Rust std library][std-float]:
//!
//!     - `sqrt`, `abs`
//!     - `exp`, `ln`
//!     - `sin`, `cos`, `tan`, `asin`, `acos`, `atan`, `atan2`
//!     - `sinh`, `cosh`, `tanh`, `asinh`, `acosh`, `atanh`
//!     - `floor`, `ceil`, `round`
//!     - `signum`
//!
//! - other functions:
//!
//!     - `max(x, ...)`, `min(x, ...)`: maximum and minimumum of 1 or more numbers
//!
//! - constants:
//!
//!     - `pi`
//!     - `e`
//!
//! # Deserialization
//!
//! [`Expr`][Expr] supports deserialization using the [serde] library to make flexible
//! configuration easy to set up, if the feature `serde` is enabled (disable by default).
//!
#![cfg_attr(feature = "serde", doc = " ```rust")]
#![cfg_attr(not(feature = "serde"), doc = " ```rust,ignore")]
//! #[macro_use]
//! extern crate toml;
//! extern crate meval;
//! extern crate serde;
//! use serde::Deserialize;
//! use meval::{Expr, Context};
//!
//! #[derive(Deserialize)]
//! struct Ode {
//!     #[serde(deserialize_with = "meval::de::as_f64")]
//!     x0: f64,
//!     #[serde(deserialize_with = "meval::de::as_f64")]
//!     t0: f64,
//!     f: Expr,
//!     g: Expr,
//! }
//!
//! fn main() {
//!     let config = r#"
//!         x0 = "cos(1.)"
//!         t0 = 2
//!         f = "sin(x)"
//!         g = 2.5
//!     "#;
//!     let ode: Ode = toml::from_str(config).unwrap();
//!
//!     assert_eq!(ode.x0, 1f64.cos());
//!     assert_eq!(ode.t0, 2f64);
//!     assert_eq!(ode.f.bind("x").unwrap()(2.), 2f64.sin());
//!     assert_eq!(ode.g.eval().unwrap(), 2.5f64);
//! }
//!
//! ```
//!
//! # Related projects
//!
//! This is a toy project of mine for learning Rust, and to be hopefully useful when writing
//! command line scripts. There is no plan to make this anything more than _math expression ->
//! number_ "converter". For more advanced scripting, see:
//!
//! - [dyon] -- A rusty dynamically typed scripting language
//! - [gluon] -- A static, type inferred programming language for application embedding
//! - [rodolf0/tox](https://github.com/rodolf0/tox) -- another shunting yard expression parser
//!
//! [Rust]: https://www.rust-lang.org/
//! [std-float]: http://doc.rust-lang.org/stable/std/primitive.f64.html
//!
//! [Expr]: struct.Expr.html
//! [Expr::bind]: struct.Expr.html#method.bind
//! [Context]: struct.Context.html
//! [serde]: https://crates.io/crates/serde
//! [dyon]: https://crates.io/crates/dyon
//! [gluon]: https://crates.io/crates/gluon

#[macro_use]
extern crate nom;
extern crate fnv;
#[cfg(feature = "serde")]
extern crate serde;
#[cfg(feature = "serde")]
use serde::{Serialize, Deserialize};
#[cfg(test)]
extern crate serde_json;
#[cfg(test)]
extern crate serde_test;

use std::fmt;

mod expr;
mod extra_math;
pub mod shunting_yard;
pub mod tokenizer;

#[cfg(feature = "serde")]
pub mod de;

pub use expr::*;
pub use shunting_yard::RPNError;
pub use tokenizer::ParseError;

/// An error produced during parsing or evaluation.
#[derive(Debug, Clone, PartialEq)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub enum Error {
    UnknownVariable(String),
    Function(String, FuncEvalError),
    /// An error returned by the parser.
    ParseError(ParseError),
    /// The shunting-yard algorithm returned an error.
    RPNError(RPNError),
    // A catch all for all other errors during evaluation
    EvalError(EvalError),
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match *self {
            Error::UnknownVariable(ref name) => {
                write!(f, "Evaluation error: unknown variable `{}`.", name)
            }
            Error::Function(ref name, ref e) => {
                write!(f, "Evaluation error: function `{}`: {}", name, e)
            }
            Error::ParseError(ref e) => {
                write!(f, "Parse error: ")?;
                e.fmt(f)
            }
            Error::RPNError(ref e) => {
                write!(f, "RPN error: ")?;
                e.fmt(f)
            }
            Error::EvalError(ref e) => {
                write!(f, "Eval error: ")?;
                e.fmt(f)
            }
        }
    }
}

impl From<ParseError> for Error {
    fn from(err: ParseError) -> Error {
        Error::ParseError(err)
    }
}

impl From<RPNError> for Error {
    fn from(err: RPNError) -> Error {
        Error::RPNError(err)
    }
}

impl std::error::Error for Error {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match *self {
            Error::ParseError(ref e) => Some(e),
            Error::RPNError(ref e) => Some(e),
            Error::Function(_, ref e) => Some(e),
            _ => None,
        }
    }
}
